processing/src/main/java/org/apache/druid/segment/nested/VariantColumnSerializer.java

```diff
@@ -299,155 +301,246 @@ public void serialize(ColumnValueSelector<? extends StructuredData> selector) th
-                            + dictionaryIdLookup.getLongCardinality()
-                            + dictionaryIdLookup.getDoubleCardinality()
-                            + dictionaryIdLookup.getArrayCardinality();
-    final CompressionStrategy compression = indexSpec.getDimensionCompression();
-    final CompressionStrategy compressionToUse;
-    if (compression != CompressionStrategy.UNCOMPRESSED && compression != CompressionStrategy.NONE) {
-      compressionToUse = compression;
-    } else {
-      compressionToUse = CompressionStrategy.LZ4;
-    }
-
-    final SingleValueColumnarIntsSerializer encodedValueSerializer = CompressedVSizeColumnarIntsSerializer.create(
-        name,
-        segmentWriteOutMedium,
-        filenameBase,
-        cardinality,
-        compressionToUse,
-        segmentWriteOutMedium.getCloser()
-    );
-    encodedValueSerializer.open();
+    closeForWrite();
+    internalSerializer.writeTo(channel, smoosher);
+  }
 
-    final GenericIndexedWriter<ImmutableBitmap> bitmapIndexWriter = new GenericIndexedWriter<>(
-        segmentWriteOutMedium,
-        name,
-        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
-    );
-    bitmapIndexWriter.open();
-    bitmapIndexWriter.setObjectsNotSorted();
-    final MutableBitmap[] bitmaps = new MutableBitmap[cardinality];
-    final Int2ObjectRBTreeMap<MutableBitmap> arrayElements = new Int2ObjectRBTreeMap<>();
-    for (int i = 0; i < bitmaps.length; i++) {
-      bitmaps[i] = indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap();
-    }
-    final GenericIndexedWriter<ImmutableBitmap> arrayElementIndexWriter = new GenericIndexedWriter<>(
-        segmentWriteOutMedium,
-        name + "_arrays",
-        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
-    );
-    arrayElementIndexWriter.open();
-    arrayElementIndexWriter.setObjectsNotSorted();
-
-    final IntIterator rows = intermediateValueWriter.getIterator();
-    int rowCount = 0;
-    final int arrayBaseId = dictionaryIdLookup.getStringCardinality()
-                            + dictionaryIdLookup.getLongCardinality()
-                            + dictionaryIdLookup.getDoubleCardinality();
-    while (rows.hasNext()) {
-      final int dictId = rows.nextInt();
-      encodedValueSerializer.addValue(dictId);
-      bitmaps[dictId].add(rowCount);
-      if (dictId >= arrayBaseId) {
-        int[] array = dictionaryIdLookup.getArrayValue(dictId);
-        for (int elementId : array) {
-          arrayElements.computeIfAbsent(
-              elementId,
-              (id) -> indexSpec.getBitmapSerdeFactory().getBitmapFactory().makeEmptyMutableBitmap()
-          ).add(rowCount);
+  /**
+   * Internal serializer used to serailize a {@link VariantColumn}. Contains the logic to write out the column to a
+   * {@link FileSmoosher}. Created by {@link VariantColumnSerializer} once it is closed for writes.
+   */
+  public static class InternalSerializer implements Serializer
```

Comment:
it still isn't completely clear why this needs to exist from the javadocs. I _think_ the reason is related to the intermediate column writer stuff (which I don't actually think needs to exist either), but it isn't clear from the javadoc

Reply:
It's essentially to be reused, since it contains only the code to write out the column. I've added that detail as well in the PR, is there any other details that should be included in the javadoc?